keps/sig-apps/4443-configurable-pod-failure-policy-reasons/README.md

```diff
@@ -0,0 +1,971 @@
+          - name: "ExitCode2"  # Will be added as a suffix to the reason "PodFailurePolicy" condition reason.
+            action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [2]
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+```
+
+#### Story 2
+
+As a user, I am using a JobSet to manage a group of jobs, each running a HPC simulation.
+Each job runs a simulation with different random initial parameters. When a simulation ends, the
+application will exit with one of two exit codes:
+
+- Exit code 2, which indicates the simulation produced an invalid result due to bad starting parameters, and should
+not be retried.
+- Exit code 3, which indicates the simulation produced an invalid result but the initial parameters were reasonable,
+so the simulation should be restarted.
+
+When a Job fails due to a pod failing with exit code 2, I want my job management software to leave the Job in
+a failed state.
+
+When a Job fails due to a pod failing with exit code 3, I want my job management software to to restart the Job.
+
+**Example JobSet for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: restart-job-example
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: {{topologyDomain}} # 1:1 job replica to topology domain assignment
+spec:
+  failurePolicy:
+    rules:
+    # If Job fails due to a pod failing with exit code 2, leave it in a failed state.
+    - action: FailJob
+      targetReplicatedJobs:
+      - simulations
+      onJobFailureReasons:
+      - PodFailurePolicy_ExitCode2  # Job failure reason format: PodFailurePolicy_{ruleName}
+    # If Job fails due to a pod failing with exit code 3, restart that Job.
+    - action: RestartJob
+      targetReplicatedJobs:
+      - simulations
+      onJobFailureReasons:
+      - PodFailurePolicy_ExitCode3  # Job failure reason format: PodFailurePolicy_{ruleName}
+    maxRestarts: 10
+  replicatedJobs:
+  - name: simulations
+    replicas: 10
+    template:
+      spec:
+        parallelism: 1
+        completions: 1
+        backoffLimit: 0
+        # Pod failure policy rules, the names of which are referenced in the JobSet failure policy.
+        podFailurePolicy:
+          rules:
+          - name: ExitCode2
+            action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [2]
+          - name: ExitCode3
+            action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [3]
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+```
+
```

Comment:
Can we get another story on PodFailurePolicy conditions? Exit code is one way for failures but we also allow for failing based on conditions. It would be worth stating how that would work also.

Reply:
The pattern is the same, what would we learn from another story?

Reply:
Mostly calling out that we want to support both and making sure we have integration tests for that also.

Reply:
I guess that it comes from name so it is probably not necessary as a user story but we should make sure to have integration/unit tests for conditions as well as exitcodes.

Reply:
Mentioned testing onPodConditions and onExitCodes in testing sections